### PR TITLE
Switch to Java 21

### PR DESCRIPTION
**Pull request by camsaul**
- Upgrade to Snowflake 3.20.0
- Add `--add-opens` option everywhere we run `java -jar` or document running it
- Bump Docker base image from Alpine+Temurin 11 to Alpine+Temurin 21
- Make sure CI runs tests against Java 21
- Bump GH Actions release stuff to build/release with Java 21

https://metaboat.slack.com/archives/C05MPF0TM3L/p1729188909680859?thread_ts=1729074092.396789&cid=C05MPF0TM3L

### CODE REVIEW

**Review by dpsutton — APPROVED** [review]

**Comment by camsaul** [comment]
Let's not merge this in yet we need to update all the Docker images and what not still

File: Dockerfile [hunk 1]
```diff
@@ -32,7 +32,7 @@ RUN INTERACTIVE=false CI=true MB_EDITION=$MB_EDITION bin/build.sh :version ${VER
 ## jar from the previous stage rather than the local build
 ## we're not yet there to provide an ARM runner till https://github.com/adoptium/adoptium/issues/96 is ready
 
-FROM --platform=linux/amd64 eclipse-temurin:11-jre-alpine as runner
+FROM --platform=linux/amd64 eclipse-temurin:21-jre-alpine as runner
```
Comment (paoliniluis) [inline Dockerfile:35]: Just a comment: as soon as we release our first ARM image, we will be able to remove the platform specification from both Dockerfiles so the docker engine can start pulling the correct one based on host OS cc @iethree

Comment (camsaul) [reply Dockerfile:35]: Fixed by #48910

**Comment by camsaul** [comment]
See also #49430 which updated the Oracle driver

**Pull request merged**